File: server/internal/model/character_card.go
package model

import "time"

const (
	CharacterCardStatusDraft     = "draft"
	CharacterCardStatusPublished = "published"

	CharacterCardVisibilityPrivate = "private"
	CharacterCardVisibilityPublic  = "public"

	CharacterCardReviewNone     = "none"
	CharacterCardReviewPending  = "pending"
	CharacterCardReviewApproved = "approved"
	CharacterCardReviewRejected = "rejected"
)

// CharacterCard is an RPBox-native character profile used for authoring and
// public presentation. It deliberately has no one-to-one constraint with the
// story archive Character model.
type CharacterCard struct {
	ID     uint `gorm:"primarykey" json:"id"`
	UserID uint `gorm:"index;not null" json:"user_id"`

	CharacterID     *uint  `gorm:"index" json:"character_id"`
	SourceBackupID  *uint  `gorm:"index" json:"source_backup_id"`
	SourceAccountID string `gorm:"size:32" json:"source_account_id"`
	SourceProfileID string `gorm:"size:128" json:"source_profile_id"`

	FirstName          string `gorm:"size:128" json:"first_name"`
	LastName           string `gorm:"size:128" json:"last_name"`
	DisplayName        string `gorm:"size:256" json:"display_name"`
	Title              string `gorm:"size:128" json:"title"`
	FullTitle          string `gorm:"size:256" json:"full_title"`
	Race               string `gorm:"size:64" json:"race"`
	Class              string `gorm:"size:64" json:"class"`
	EyeColor           string `gorm:"size:64" json:"eye_color"`
	EyeColorHex        string `gorm:"size:16" json:"eye_color_hex"`
	Age                string `gorm:"size:64" json:"age"`
	Height             string `gorm:"size:64" json:"height"`
	Weight             string `gorm:"size:64" json:"weight"`
	Birthplace         string `gorm:"size:256" json:"birthplace"`
	Residence          string `gorm:"size:256" json:"residence"`
	RelationshipStatus string `gorm:"size:64" json:"relationship_status"`
	Icon               string `gorm:"size:128" json:"icon"`
	ClassColor         string `gorm:"size:16" json:"class_color"`
	NameColor          string `gorm:"size:16" json:"name_color"`

	Summary         string `gorm:"size:1000" json:"summary"`
	BackgroundStory string `gorm:"type:text" json:"background_story"`
	// FirstImpression is the legacy rich-text field retained as free-form
	// "other notes" beside the structured five-slot impressions.
	FirstImpression string `gorm:"type:text" json:"first_impression"`
	OtherContent    string `gorm:"type:text" json:"other_content"`

	// PortraitImage is an internal storage reference. API responses expose only
	// the permission-checked image endpoint built by the CharacterCard DTO.
	PortraitImage          string     `gorm:"size:512" json:"-"`
	PortraitImageUpdatedAt *time.Time `json:"portrait_image_updated_at"`

	Status     string `gorm:"size:20;not null;default:draft;index" json:"status"`
	Visibility string `gorm:"size:20;not null;default:private;index" json:"visibility"`
	SortOrder  int    `gorm:"default:0;index" json:"sort_order"`

	ReviewStatus  string     `gorm:"size:20;not null;default:none;index" json:"review_status"`
	ReviewerID    *uint      `gorm:"index" json:"reviewer_id,omitempty"`
	ReviewComment string     `gorm:"size:512" json:"review_comment"`
	ReviewedAt    *time.Time `json:"reviewed_at,omitempty"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// CharacterCardPortrait stores the ordered protected portrait gallery. The
// first row by SortOrder is the cover and is mirrored to CharacterCard's
// legacy PortraitImage fields for older clients and stable cover URLs.
type CharacterCardPortrait struct {
	ID              uint `gorm:"primarykey" json:"id"`
	CharacterCardID uint `gorm:"not null;index;uniqueIndex:idx_character_card_portraits_card_sort" json:"character_card_id"`
	SortOrder       int  `gorm:"not null;uniqueIndex:idx_character_card_portraits_card_sort" json:"sort_order"`

	Image          string     `gorm:"size:512;not null" json:"-"`
	ImageUpdatedAt *time.Time `json:"image_updated_at"`

	CharacterCard CharacterCard `gorm:"foreignKey:CharacterCardID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"-"`
	CreatedAt     time.Time     `json:"created_at"`
	UpdatedAt     time.Time     `json:"updated_at"`
}

// CharacterCardPublication is the last moderator-approved immutable view of a
// card. Owners edit the live aggregate while public readers stay on this
// snapshot until the next approval, preventing unreviewed text or images from
// leaking through an already-public card.
type CharacterCardPublication struct {
	CharacterCardID uint   `gorm:"primarykey" json:"character_card_id"`
	UserID          uint   `gorm:"not null;index" json:"user_id"`
	Payload         string `gorm:"type:text;not null" json:"-"`

	ApprovedBy *uint     `gorm:"index" json:"approved_by,omitempty"`
	ApprovedAt time.Time `gorm:"not null" json:"approved_at"`
	CreatedAt  time.Time `json:"created_at"`
	UpdatedAt  time.Time `json:"updated_at"`
}

// CharacterCardSubmission is the single latest owner-submitted snapshot that
// moderators review. The editable CharacterCard aggregate may continue to
// auto-save after submission without mutating this frozen candidate. A later
// publish action replaces this row instead of creating another queue item.
type CharacterCardSubmission struct {
	CharacterCardID uint   `gorm:"primarykey" json:"character_card_id"`
	UserID          uint   `gorm:"not null;index" json:"user_id"`
	Payload         string `gorm:"type:text;not null" json:"-"`

	SubmittedAt time.Time `gorm:"not null;index" json:"submitted_at"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// CharacterCardImpression is one of the five fixed "at first glance" slots
// attached to a character card. Image fields are private storage references;
// API responses expose them only through permission-aware image endpoints.
type CharacterCardImpression struct {
	ID              uint  `gorm:"primarykey" json:"id"`
	CharacterCardID uint  `gorm:"not null;uniqueIndex:idx_character_card_impressions_card_slot" json:"character_card_id"`
	Slot            uint8 `gorm:"not null;uniqueIndex:idx_character_card_impressions_card_slot;check:slot >= 1 AND slot <= 5" json:"slot"`

	Active   bool   `gorm:"not null;default:false" json:"active"`
	Title    string `gorm:"size:80" json:"title"`
	Text     string `gorm:"size:500" json:"text"`
	TRP3Icon string `gorm:"size:128" json:"trp3_icon"`

	IconImage          string     `gorm:"size:512" json:"-"`
	IconImageUpdatedAt *time.Time `json:"icon_image_updated_at"`
	Image              string     `gorm:"size:512" json:"-"`
	ImageUpdatedAt     *time.Time `json:"image_updated_at"`

	CharacterCard CharacterCard `gorm:"foreignKey:CharacterCardID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"-"`
	CreatedAt     time.Time     `json:"created_at"`
	UpdatedAt     time.Time     `json:"updated_at"`
}

File: mobile/src/api/characterCard.ts
import { request } from '@shared/api/request'

export type CharacterCardStatus = 'draft' | 'published'
export type CharacterCardVisibility = 'private' | 'public'
export type CharacterCardReviewStatus = 'none' | 'pending' | 'approved' | 'rejected'
export type CharacterCardSourceType = 'blank' | 'backup'

export interface CharacterCardPortrait {
  id: number
  image_url: string
  image_updated_at?: string | null
  sort_order: number
  is_cover: boolean
}

export type CharacterCardPortraitImage = CharacterCardPortrait

export interface CharacterCardImpressionUpdate {
  slot: number
  active: boolean
  title: string
  text: string
  trp3_icon: string
  icon_image_url: string
  image_url: string
}

export interface CharacterCardImpression extends CharacterCardImpressionUpdate {
  icon_image_updated_at?: string | null
  image_updated_at?: string | null
}

export interface CharacterCardTRP3Color {
  r: number
  g: number
  b: number
}

export interface CharacterCardAdditionalInfo {
  id: number
  name: string
  value: string
  icon: string
}

export interface CharacterCardPersonalityTrait {
  preset_id: number | null
  left_text: string
  right_text: string
  left_icon: string
  right_icon: string
  left_color: CharacterCardTRP3Color | null
  right_color: CharacterCardTRP3Color | null
  value: number
}

export interface CharacterCardSource {
  backup_id: number
  account_id: string
  profile_id: string
  profile_name?: string
  display_name?: string
  first_name?: string
  last_name?: string
  title?: string
  race?: string
  class?: string
  icon?: string
  class_color?: string
  name_color?: string
  backup_updated_at?: string
}

export interface CharacterCardSummary {
  id: number
  user_id: number
  character_id?: number | null
  source_backup_id?: number | null
  source_account_id?: string | null
  source_profile_id?: string | null
  first_name: string
  last_name: string
  display_name: string
  title: string
  full_title: string
  race: string
  class: string
  eye_color: string
  eye_color_hex: string
  age: string
  height: string
  weight: string
  birthplace: string
  residence: string
  relationship_status: string
  icon?: string
  class_color: string
  name_color: string
  additional_info?: CharacterCardAdditionalInfo[]
  personality_traits?: CharacterCardPersonalityTrait[]
  summary: string
  portrait_image_url?: string | null
  portrait_image_updated_at?: string | null
  portraits?: CharacterCardPortrait[]
  impressions?: CharacterCardImpression[]
  status: CharacterCardStatus
  visibility: CharacterCardVisibility
  review_status?: CharacterCardReviewStatus | null
  review_comment?: string | null
  reviewed_at?: string | null
  sort_order?: number
  created_at?: string
  updated_at: string
}

export interface CharacterCard extends CharacterCardSummary {
  background_story: string
  first_impression: string
  impressions: CharacterCardImpression[]
  other_content: string
}

export interface CharacterCardShare {
  path: string
  title: string
  summary: string
  updated_at: string
}

export interface CreateCharacterCardRequest {
  source_type: CharacterCardSourceType
  source_backup_id?: number
  source_profile_id?: string
}

export interface UpdateCharacterCardRequest {
  character_id?: number | null
  first_name: string
  last_name: string
  display_name: string
  title: string
  full_title: string
  race: string
  class: string
  eye_color: string
  eye_color_hex: string
  age: string
  height: string
  weight: string
  birthplace: string
  residence: string
  relationship_status: string
  icon: string
  class_color: string
  name_color: string
  additional_info: CharacterCardAdditionalInfo[]
  personality_traits: CharacterCardPersonalityTrait[]
  summary: string
  background_story: string
  first_impression: string
  impressions: CharacterCardImpressionUpdate[]
  other_content: string
  portrait_image_url: string
  status: CharacterCardStatus
  visibility: CharacterCardVisibility
  sort_order?: number
}

interface DataEnvelope<T> {
  data?: T
}

function unwrapData<T>(response: T | DataEnvelope<T>): T {
  if (response && typeof response === 'object' && 'data' in response) {
    const data = (response as DataEnvelope<T>).data
    if (data !== undefined) return data
  }
  return response as T
}

function unwrapCharacterCard(response: unknown): CharacterCard {
  const body = unwrapData(
    response as CharacterCard | DataEnvelope<CharacterCard | { character_card: CharacterCard }>,
  )
  if (body && typeof body === 'object' && 'character_card' in body) {
    return (body as { character_card: CharacterCard }).character_card
  }
  return body as CharacterCard
}

export function isApprovedPublicCharacterCard(
  card: Pick<CharacterCardSummary, 'status' | 'visibility' | 'review_status'> | null | undefined,
) {
  return card?.status === 'published'
    && card.visibility === 'public'
    && card.review_status === 'approved'
}

export async function getCharacterCardSources(): Promise<{ sources: CharacterCardSource[] }> {
  const response = await request.get<
    { sources: CharacterCardSource[] } | DataEnvelope<{ sources: CharacterCardSource[] }>
  >('/character-card-sources')
  const body = unwrapData(response)
  return { sources: Array.isArray(body?.sources) ? body.sources : [] }
}

export async function listMyCharacterCards(): Promise<{ character_cards: CharacterCardSummary[] }> {
  const response = await request.get<
    { character_cards: CharacterCardSummary[] }
    | DataEnvelope<{ character_cards: CharacterCardSummary[] }>
  >('/character-cards')
  const body = unwrapData(response)
  return {
    character_cards: Array.isArray(body?.character_cards) ? body.character_cards : [],
  }
}

export async function createCharacterCard(payload: CreateCharacterCardRequest): Promise<CharacterCard> {
  const response = await request.post<unknown>('/character-cards', payload)
  return unwrapCharacterCard(response)
}

export async function getCharacterCard(id: number): Promise<CharacterCard> {
  const response = await request.get<unknown>(`/character-cards/${id}`)
  return unwrapCharacterCard(response)
}

export async function getCharacterCardShare(id: number): Promise<CharacterCardShare> {
  const response = await request.get<CharacterCardShare | DataEnvelope<CharacterCardShare>>(
    `/character-cards/${id}/share`,
  )
  const body = unwrapData(response)
  if (!body?.path) throw new Error('Missing character card share path')
  return body
}

export async function updateCharacterCard(
  id: number,
  payload: Partial<UpdateCharacterCardRequest>,
): Promise<CharacterCard> {
  const response = await request.put<unknown>(`/character-cards/${id}`, payload)
  return unwrapCharacterCard(response)
}

export async function publishCharacterCard(id: number): Promise<CharacterCard> {
  const response = await request.post<unknown>(`/character-cards/${id}/publish`)
  return unwrapCharacterCard(response)
}

export async function deleteCharacterCard(id: number): Promise<void> {
  await request.delete(`/character-cards/${id}`)
}

export async function uploadCharacterCardPortrait(file: File): Promise<string> {
  const formData = new FormData()
  formData.append('image', file)
  const response = await request.post<
    { portrait_image_ref?: string } | DataEnvelope<{ portrait_image_ref?: string }>
  >('/upload/character-card-portrait', formData)
  const body = unwrapData(response)
  const portraitRef = body?.portrait_image_ref?.trim() || ''
  if (!portraitRef) throw new Error('Missing character-card portrait reference')
  return portraitRef
}

export async function addCharacterCardPortrait(id: number, imageRef: string): Promise<CharacterCard> {
  const response = await request.post<unknown>(`/character-cards/${id}/portraits`, {
    image_ref: imageRef,
  })
  return unwrapCharacterCard(response)
}

export async function setCharacterCardPortraitCover(
  id: number,
  portraitId: number,
): Promise<CharacterCard> {
  const response = await request.put<unknown>(`/character-cards/${id}/portraits/${portraitId}/cover`)
  return unwrapCharacterCard(response)
}

export async function deleteCharacterCardPortrait(
  id: number,
  portraitId: number,
): Promise<CharacterCard> {
  const response = await request.delete<unknown>(`/character-cards/${id}/portraits/${portraitId}`)
  return unwrapCharacterCard(response)
}
